### PR TITLE
wgengine/netstack: only accept connection after dialing

### PR DESCRIPTION
**Pull request by andrew-d**
If we accept a forwarded TCP connection before dialing, we can
erroneously signal to a client that we support IPv6 (or IPv4) without
that actually being possible. Instead, we only complete the client's TCP
handshake after we've dialed the outbound connection; if that fails, we
respond with a RST.

Signed-off-by: Andrew Dunham <andrew@tailscale.com>

### CODE REVIEW

File: wgengine/netstack/netstack.go [hunk 1]
```diff
@@ -743,86 +743,119 @@ func (ns *Impl) acceptTCP(r *tcp.ForwarderRequest) {
...
-	ns.forwardTCP(c, clientRemoteIP, &wq, dialAddr)
+
+	if !ns.forwardTCP(createConn, clientRemoteIP, &wq, dialAddr) {
+		r.Complete(true) // sends a RST
+	}
 }
 
-func (ns *Impl) forwardTCP(client *gonet.TCPConn, clientRemoteIP netip.Addr, wq *waiter.Queue, dialAddr netip.AddrPort) {
-	defer client.Close()
+func (ns *Impl) forwardTCP(getClient func() *gonet.TCPConn, clientRemoteIP netip.Addr, wq *waiter.Queue, dialAddr netip.AddrPort) (handled bool) {
```
Comment (bradfitz) [inline wgengine/netstack/netstack.go:850]: maybe worth making a named type for that func. quite the signature.

File: wgengine/netstack/netstack.go [hunk 2]
```diff
@@ -840,13 +873,29 @@ func (ns *Impl) forwardTCP(client *gonet.TCPConn, clientRemoteIP netip.Addr, wq
...
+	// respond to the client with a RST. Either way, the caller no longer
+	// needs to clean up the client connection.
+	handled = true
+
+	// We dialed the connection; we can complete the client's TCP handshake.
+	client := getClient()
+	if client == nil {
+		return
+	}
+	defer client.Close()
```
Comment (bradfitz) [inline wgengine/netstack/netstack.go:897]: where was this before?

**Review by bradfitz — APPROVED** [review]
LGTM assuming it works!

**Review by crawshaw — APPROVED** [review]
Conceptually looks good! Not an easy thing to write a test for. :)

File: wgengine/netstack/netstack.go (hunk 2 above)
Comment (andrew-d) [reply wgengine/netstack/netstack.go:897]: At the top of `forwardTCP`

File: wgengine/netstack/netstack.go (hunk 1 above)
Comment (andrew-d) [reply wgengine/netstack/netstack.go:850]: Something like `type GetClientFunc func() *gonet.TCPConn`? Or do you mean passing all the arguments (`getClient`, `clientRemoteIP`, etc.) as a struct?

Comment (bradfitz) [reply wgengine/netstack/netstack.go:850]: oh, nevermind, I thought all that clientRemoteIP etc were part of the signature. nevermind me.

carry on

**Review by willnorris — APPROVED** [review]
tested and confirmed that this resolved the issue on my Synology running DSM 7.

**Comment by bradfitz** [comment]
But does actual IPv6 still work? 😜

**Comment by mihaip** [comment]
Should this reference #5425?

**Comment by bradfitz** [comment]
> Should this reference #5425?

I often write footer lines like `Updates #nnn (maybe fixes!)` so it's not recognized by GitHub regexps and auto-closed, lest it doesn't fix it.

**Comment by DentonGentry** [comment]
This would be fine to go in now, `release-branch/1.30` was created this morning.

**Comment by andrew-d** [comment]
I launched two servers in Digital Ocean, with IPv6 enabled on both, and confirmed that IPv6 access through the exit node works fine (using `curl -6 https://ifconfig.co`). Given that, and with @willnorris reporting that it fixes things on his Synology, I'm going to merge this now.

**Pull request merged**